File: test/system/auto/stress/metadataMaxFiles.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from TestUtils import TestUtilsMixin
import os

import logging
log = logging.getLogger('test.auto')

class MetadataMaxFiles(TestUtilsMixin, unittest.TestCase):
    "open a large !METADATA with too few files"

    order=75

    settings = TestUtilsMixin.settings.copy()
    settings['tserver.compaction.major.delay'] = 1

    def runTest(self):
        # Create a bunch of tables with splits to split the !METADATA table
        self.splitfile = 'splits'
        fp = open(self.splitfile, 'w')
        for i in range(1000):
            fp.write('%03d\n' % i)
        fp.close()
        self.splitfile = os.path.realpath(self.splitfile)
        self.shell(self.masterHost(),
                   'config -t !METADATA -s table.split.threshold=10000\n' + 
                   ''.join(['createtable test%d -sf %s\nflush -t !METADATA\n' % (i, self.splitfile) for i in range(5)]))
        self.shutdown_accumulo(150)
        
        # reconfigure accumulo to use a very small number of files
        self.stop_accumulo()
        self.settings['tserver.scan.files.open.max'] = 10
        self.create_config_file(self.settings)

        # make sure the master knows about all the tables we created
        self.sleep(5)
        self.start_accumulo()
        self.sleep(60)
        h = self.runOn(self.masterHost(),
                       [self.accumulo_sh(), 'org.apache.accumulo.server.test.GetMasterStats'])
        out, err = h.communicate()
        self.assert_(len([x for x in out.split('\n') if x.find('  Tablets 1001') == 0]) == 5)

    def tearDown(self):
        TestUtilsMixin.tearDown(self)
        os.unlink(self.splitfile)

def suite():
    result = unittest.TestSuite()
    result.addTest(MetadataMaxFiles())
    return result

File: test/src/test/java/org/apache/accumulo/test/functional/MetadataMaxFiles.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.accumulo.test.functional;

import static org.junit.Assert.assertEquals;

import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.SortedSet;
import java.util.TreeSet;

import org.apache.accumulo.core.client.Connector;
import org.apache.accumulo.core.client.impl.MasterClient;
import org.apache.accumulo.core.client.security.tokens.PasswordToken;
import org.apache.accumulo.core.conf.Property;
import org.apache.accumulo.core.master.thrift.MasterClientService.Client;
import org.apache.accumulo.core.master.thrift.MasterMonitorInfo;
import org.apache.accumulo.core.master.thrift.TableInfo;
import org.apache.accumulo.core.master.thrift.TabletServerStatus;
import org.apache.accumulo.core.metadata.MetadataTable;
import org.apache.accumulo.core.metadata.RootTable;
import org.apache.accumulo.core.security.CredentialHelper;
import org.apache.accumulo.core.security.thrift.TCredentials;
import org.apache.accumulo.core.util.UtilWaitThread;
import org.apache.accumulo.minicluster.MiniAccumuloConfig;
import org.apache.accumulo.server.util.Admin;
import org.apache.accumulo.trace.instrument.Tracer;
import org.apache.hadoop.io.Text;
import org.junit.Test;

public class MetadataMaxFiles extends MacTest {
  
  @Override
  public void configure(MiniAccumuloConfig cfg) {
    Map<String,String> siteConfig = new HashMap<String,String>();
    siteConfig.put(Property.TSERV_MAJC_DELAY.getKey(), "1");
    siteConfig.put(Property.TSERV_SCAN_MAX_OPENFILES.getKey(), "10");
    cfg.setSiteConfig(siteConfig);
  }
  
  @Test(timeout = 240 * 1000)
  public void test() throws Exception {
    Connector c = getConnector();
    SortedSet<Text> splits = new TreeSet<Text>();
    for (int i = 0; i < 1000; i++) {
      splits.add(new Text(String.format("%03d", i)));
    }
    c.tableOperations().setProperty(MetadataTable.NAME, Property.TABLE_SPLIT_THRESHOLD.getKey(), "10000");
    for (int i = 0; i < 5; i++) {
      String tableName = "table" + i;
      log.info("Creating " + tableName);
      c.tableOperations().create(tableName);
      log.info("adding splits");
      c.tableOperations().addSplits(tableName, splits);
      log.info("flushing");
      c.tableOperations().flush(MetadataTable.NAME, null, null, true);
      c.tableOperations().flush(RootTable.NAME, null, null, true);
    }
    UtilWaitThread.sleep(20 * 1000);
    log.info("shutting down");
    assertEquals(0, cluster.exec(Admin.class, "stopAll").waitFor());
    cluster.stop();
    log.info("starting up");
    cluster.start();
    
    UtilWaitThread.sleep(30 * 1000);
    
    while (true) {
      MasterMonitorInfo stats = null;
      TCredentials creds = CredentialHelper.create("root", new PasswordToken(MacTest.PASSWORD), c.getInstance().getInstanceName());
      Client client = null;
      try {
        client = MasterClient.getConnectionWithRetry(c.getInstance());
        stats = client.getMasterStats(Tracer.traceInfo(), creds);
      } finally {
        if (client != null)
          MasterClient.close(client);
      }
      int tablets = 0;
      for (TabletServerStatus tserver : stats.tServerInfo) {
        for (Entry<String,TableInfo> entry : tserver.tableMap.entrySet()) {
          if (entry.getKey().startsWith("!"))
            continue;
          tablets += entry.getValue().onlineTablets;
        }
      }
      if (tablets == 5005)
        break;
      UtilWaitThread.sleep(1000);
    }
  }
  
}
